pkg/storage/chunkfs/scanner.go

```diff
@@ -0,0 +1,362 @@
+		GlobalSyncTimeout:     time.Minute,
+	}
+}
+
+// String implements fmt.Stringer
+func (sc ScannerConfig) String() string {
+	b, _ := json.MarshalIndent(sc, "", "  ")
+	return string(b)
+}
+
+// Init implements linker.Initializer
+func (s *Scanner) Init(ctx context.Context) error {
+	s.logger.Infof("initializing cfg:\n%s", s.cfg)
+	go s.watcher(ctx)
+	return nil
+}
+
+func (s *Scanner) watcher(ctx context.Context) {
+	s.logger.Infof("starting watcher()")
+	defer s.logger.Infof("exiting from watcher()")
+
+	if s.cfg.SyncWorkers <= 0 {
+		s.logger.Infof("the number of SyncWorkers in the config is zero or negative, exiting from the watcher")
+		return
+	}
+
+	s.watcherWithF(ctx, s.scanDir)
+}
+
+func (s *Scanner) watcherWithF(ctx context.Context, f func(ctx context.Context, dir string)) {
+	for {
+		start := time.Now()
+		fis := files.ListDir(s.cfg.DataPath)
+		var wg sync.WaitGroup
+		sema := make(chan struct{}, s.cfg.SyncWorkers)
+		for _, fi := range fis {
+			// ignore all not-dirs
+			if !fi.IsDir() {
+				continue
+			}
+			sema <- struct{}{}
+			wg.Add(1)
+			go func(dir string) {
+				defer wg.Done()
+				defer func() {
+					<-sema
+				}()
+				f(ctx, dir)
+			}(filepath.Join(s.cfg.DataPath, fi.Name()))
+		}
+		wg.Wait()
+		diff := time.Now().Sub(start)
+		if diff < s.cfg.GlobalSyncTimeout {
+			select {
+			case <-ctx.Done():
+				// game over. Everything seemst to be stopped
+				return
+			case <-time.After(s.cfg.GlobalSyncTimeout - diff):
+				// ok, no leaks here, because it is the main case for the flow
+			}
+		}
+	}
+}
+
+func newScanInfos() scanInfos {
+	return make(scanInfos)
+}
+
+func (s *Scanner) scanDir(ctx context.Context, dir string) {
+	sis := s.buildScanInfos(dir)
+
+	sfn := filepath.Join(dir, cScanFileName)
+	if err := sis.readFromFile(sfn); err != nil {
+		s.logger.Warnf("could not read data from the file %s: %s", sfn, err)
+	}
+
+	s.sweep(ctx, dir, sis)
+	s.replicate(ctx, sis)
+	sis.writeToFile(sfn)
+}
+
+func (s *Scanner) buildScanInfos(dir string) scanInfos {
+	sis := newScanInfos()
+	for _, fi := range files.ListDir(dir) {
+		if fi.IsDir() {
+			continue
+		}
+		var si scanInfo
+		if doesLookLikeID(fi.Name()) {
+			si = scanInfo{ID: fi.Name(), Mtime: fi.ModTime(), size: fi.Size()}
+		} else if isItScanInfoFile(fi.Name()) {
+			var err error
+			sifn := filepath.Join(dir, fi.Name())
+			si, err = readScanInfo(sifn)
+			_ = os.Remove(sifn)
+			if err != nil {
+				s.logger.Warnf("could not read scanInfo from %s: %s", sifn, err)
+				continue
+			}
+		} else {
+			continue
+		}
+
+		// merge the syncInfo (si) with what we have so far about the chunk ID
+		if oldSI, ok := sis[si.ID]; ok {
+			if oldSI.Mtime.After(si.Mtime) {
+				si.Mtime = oldSI.Mtime
+			}
+			if si.SyncTime == nil {
+				si.SyncTime = oldSI.SyncTime
+			}
+			si.size = max(si.size, oldSI.size)
+		}
+		sis[si.ID] = si
+	}
+	return sis
+}
+
+func (s *Scanner) sweep(ctx context.Context, dir string, sis scanInfos) {
```

Comment:
question: can we get a lot of small files in the folder so that we don't hit the size threshold, but we will have to deal with a lot of file descriptors? Just trying to understand if `sweeping by quantity` is necessary.

Reply:
Sweeper doesn’t consume fds, we can have million of chunks and still use just few fds.

the chunks have limit on number of files to be used, see Provider.

Reply:
Got it, mainly I was more `worried` about a scenario where we need to query data where the number of files will be large while the amount of data in them will be small (thus not triggering the sweeper), that's all. Although in any case this is a kind of rare `corner` case.